### DISCUSSION TITLE
Non persistent params

### DISCUSSION THREAD

**Discussion by heroic** [expo/router#303] [Q&A] [answered]
Let's say we have an ecommerce app, and we are navigating from List of Products page to Product details page. Here the only relevant parameter to have in the URL would possibly be an `id` or `slug`. However, it may be possible to send the name, price, and image from the list of products page to the product details page, but it makes no sense to have these in the url, as they may add no value when the url is shared.

We could implement this with `localParams` or something similar, where the parameters are available to the next page, but are not added to the URL.

**Comment by EvanBacon** [comment, marked as answer]
Use [React Context](https://reactjs.org/docs/context.html) to send params around in-memory without having them live in the query parameters.

**Reply by heroic** [reply]
Maybe, I didn't get you fully, but as far as I can tell, Context will need to be cleared once consumed to avoid side-effects.

Eg: ScreenA gets the in-memory params, and user goes to ScreenB, which is part of same layout. The in-memory params will still be available to ScreenB, despite not having being passed to it.

**Reply by satya164** [reply]
Params are dependencies of the screen, like arguments to a function. A URL represents those dependencies. When you refresh a page, you'd still get the same page because the page can be reconstructed from its params, i.e. the URL. Having params that are not in the URL means the basic functionality of refreshing the page, opening the page in new tab, or sharing the URL as you mentioned, etc. no longer works. It can no longer create the same page with the same URL because there's additional information missing that's not in the URL.

Params shouldn't contain full data. It should only include the ID of the object and the screen should know how to retrieve that data from API or cache, e.g. with a data fetching library such as React Query or your own utilities.

https://reactnavigation.org/docs/params/#what-should-be-in-params

**Reply by heroic** [reply]
@satya164 I do agree, but in cases when only partial data is available, pages can be made to look like loading faster by simply sending partially available data to the next screen in the params along with the id. The screen then continues to fetch data using the id and replace the stale partial data. This makes for a better browsing experience for the customer.

Fully in agreement that this partial data should not be available in the url. However, getting this from the cache of another api request also wouldn't make sense from a keep things independent perspective.

Eg of such things would be twitter where when you click on a user's profile pic on the timeline, you instantly see their basic details, but the rest of the info is fetched after screen opens, but at the same time when you share the profile url, it first loads blank, then loads profile details and then the rest of the stuff. The url in both cases is just the user id.

**Reply by heroic** [reply]
Also to be clear, I'm not talking about adding URL parameters, which is what you have described in the link. I'm simply talking about a way to pass in props to the next screen. URLs are valid in the context of web. In a mobile app, one could argue to simply pass in whatever in the params, as long it's serializable, without affecting the side effects mentioned, again assuming that the apis are called using the id and data is updated.

On the web side of things, I'm suggesting we keep the same structure where the URL is using only id, and other props can be passed to the next screen.

Again an example of this can we with: Login with phone number and navigating to OTP. It would be better to show the number where the OTP was sent to, but having this in the URL would simply be a grave error and a security breach.

This would also make it possible to have the same url show the OTP screen which is not expected to be directly opened directly. Anyone directly opening the OTP URL should actually be seeing the login screen.

**Reply by satya164** [reply]
> Fully in agreement that this partial data should not be available in the url. However, getting this from the cache of another api request also wouldn't make sense from a keep things independent perspective.

I don't see why it wouldn't make sense. If partial data is available, it's simpler for the screen itself to read that data than trying to pass that data from every other place where navigation happens and then read from there. Alternatively, libraries like React query also allow you to update the cache of another API which is another alternative - if you're worried about one API knowing about a different API, it's not that different from one screen knowing the data another screen requires and passing that in params.

Let's take example of Twitter. On feed screen, each tweet includes the avatar of the person who posted the tweet. If you navigate from here to profile screen of that person, it's easy to pass the params. However, lets say that tweet also mentioned a handle, you have partial data for that handle, but not in this tweet - so this navigation cannot pass this data. If the screen itself reads the data then these things don't matter. Users will have a more consistent experience.

In the end, it's not really a navigation concern. It's more about if partial data is available, then how do we display that for better UX than how can we pass the data during navigation. IMO it shouldn't be the responsibility of the components that trigger navigation to this screen, then it becomes inconsistent and additional work everywhere else.

> Again an example of this can we with: Login with phone number and navigating to OTP. It would be better to show the number where the OTP was sent to, but having this in the URL would simply be a grave error and a security breach.

Don't you have a global store with such data already that comes from server (e.g. which number the OTP was sent to)? If you already have a global store/cache for API responses then this isn't a problem at all as that screen can read whatever info it needs from that store.

Also, imagine in this case you refresh the OTP page. User isn't supposed to go there directly but for sure you have refreshed such pages in the past due to some issue. Then passing info via params falls apart. If the screen read the data using something like React Query then this data will be fetched if it didn't exist and will be displayed.
